### PR TITLE
Make Python3 compatible

### PR DESCRIPTION
**Pull request by gijzelaerr**

### CODE REVIEW

**Comment by gijzelaerr** [comment]
This depends on:

casacore/casacore#853
casacore/python-casacore#159

for the improved py2 unicode support.

py3 only has strings (which are unicode), py2 has string and unicode objects.

I think by now it all works properly with py3, but for py2 sometimes a unicode object slips in, and casacore doesn't support unicode fully yet. I've been wrestling with this for a while but got a bit stuck and ran out of time, and now I'm waiting for Ger to see what he can do.

The last time I checked a basic cubical run was working with both py2 and py3, but I think you astronomers need to test it more now.

File: cubical/data_handler/__init__.py [hunk 1]
```diff
@@ -17,7 +17,7 @@ def uniquify(values):
     uniq = np.array(sorted(set(values)))
     rmap = {x: i for i, x in enumerate(uniq)}
     # apply this map to the time column to construct a timestamp column
-    indices  = np.fromiter(map(rmap.__getitem__, values), int)
+    indices  = np.fromiter(list(map(rmap.__getitem__, values)), int)
```
Comment (o-smirnov) [inline cubical/data_handler/__init__.py]: Why make a list from an iterator, only to call fromiter()?

**Comment by o-smirnov** [comment]
OK this is going to need to be updated, given the latest merge to master.

I have a proposal w.r.t. logging. I quite liked the old syntax of
```
print>>log(level[,color]), stuff
```
as it made the log statements stand out nicely, and the log level and severity was visible upfront. The py3 equivalent 
```
print(stuff, file=log(level[,color])) 
```
looks clumsier, and the log invocation is at the back, where having long "stuff" makes it harder to see. I therefore propose adding a ``print`` function to the logger class, so that the statement can be written as
```
log(level[,color]).print(stuff)
```
Reasonable enough?

**Comment by bennahugo** [comment]
how about the more traditional log.info, log.warn, log.error?

**Comment by cyriltasse** [comment]
yeah - loooks good to me!

**Comment by o-smirnov** [comment]
Well the current logger is just a small wrapper around the standard logging functions... ``log.print(stuff)`` is just ``logger.info(stuff)`` underneath. But I like to have finer-grained control over the levels (and colors), hence the wrapper.

``log(2, "red").print(stuff)`` is the equivalent of ``logger.log(logger.INFO-2, ModColor.Str(stuff, "red"))``, but reads a lot better IMHO. And knows how to turn off colors when logging stuff to the logfile.

**Comment by o-smirnov** [comment]
> yeah - loooks good to me!

Yep, I have backporting this into DDF in mind as well.

**Comment by gijzelaerr** [comment]
Why not just use the standard logging facilities, then parts of cubical can also be integrated with other tools that use the standard logging. It is easy to add a custom log level, and you can add a helper log function per custom log level.

I'm not sure what you mean by that it is aware of colors and stuff, those are things you should configure in your logging handler and formatter.

I'm not sure what you mean by

**Comment by o-smirnov** [comment]
> Why not just use the standard logging facilities, then parts of cubical can also be integrated with other 
> tools that use the standard logging. 

That's exactly what it's doing already... It's a small wrapper around the standard facility.

**Comment by sjperkins** [comment]
So this is something I got wrong a lot, and I'd recommend the standard pythonic way within standard library code (i.e. Cubical routines).

```python
import logging

log = logging.getLogger(__name__)

def astro_function(*args, **kwargs):
   log.info("Informational")
   log.warn("Warning")
   log.error("Error")
```

Then, when you create an actual application (like gocubical), you create an actual logger with formatting, different handlers, levels etc. The following is lifted from `tricolour`

```python
def create_logger():
    """ Create a console logger """
    log = logging.getLogger("cubical")
    cfmt = logging.Formatter(
        ('%(name)s - %(asctime)s %(levelname)s - %(message)s'))
    log.setLevel(logging.DEBUG)
    filehandler = logging.FileHandler("cubical.log")
    filehandler.setFormatter(cfmt)
    log.addHandler(filehandler)
    log.setLevel(logging.INFO)

    console = logging.StreamHandler()
    console.setLevel(logging.INFO)
    console.setFormatter(cfmt)

    log.addHandler(console)

    return log, filehandler, console, cfmt
```

**Comment by o-smirnov** [comment]
That's pretty much exactly what ``cubical/tools/logger.py`` does...

All I want is finer-grained message levels without making the code look ugly! And colors. Can't I keep a wrapper around the standard Pythonic library? Pretty professorial please?

**Comment by sjperkins** [comment]
> All I want is finer-grained message levels without making the code look ugly! And colors. Can't I keep a wrapper around the standard Pythonic library? Pretty professorial please?

Yeah go for it, it might look a bit strange to someone eyeballing the codebase for the first time, but colours are important ;-)

**Comment by o-smirnov** [comment]
> but colours are important ;-)

Exactly! I don't want to reinvent the wheel. I just want shinier rims.

**Comment by o-smirnov** [comment]
As to someone new eyeballing the code, if they can't figure out for themselves what ``log.print()`` is meant to do, do we really want them eyeballing the code in the first place?

**Comment by sjperkins** [comment]
> As to someone new eyeballing the code, if they can't figure out for themselves what log.print() is meant to do, do we really want them eyeballing the code in the first place?

They'll probably figure it out :-)

**Comment by gijzelaerr** [comment]
are you sure wanted to close this PR @o-smirnov

**Comment by o-smirnov** [comment]
No that was a misclick.

**Comment by o-smirnov** [comment]
Just for the record, the following regex ``print>>\s*(([^\(]+)(\([^\(]+\))?),\s*(.*)`` replaced by ``$1.print($4)`` achieves the desired effect.

**Comment by o-smirnov** [comment]
Unless the print statement was multiline, but there are few enough of those...

**Comment by gijzelaerr** [comment]
ah well, i guess I should just start over with the py3 port.

**Comment by gijzelaerr** [comment]
this has been redone in https://github.com/ratt-ru/CubiCal/pull/270

**Pull request closed without merging**